src/dbnode/persist/fs/migration/migration.go

```diff
@@ -0,0 +1,101 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package migration
+
+import (
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/persist"
+	"github.com/m3db/m3/src/dbnode/persist/fs"
+	"github.com/m3db/m3/src/dbnode/storage"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+// Migration interface is implemented by tasks that wish to perform a data migration
+// on a fileset. Typically involves updating files in a fileset that were created by
+// a previous version of the database client
+type Migration interface {
+
+	// Run is the set of steps to successfully complete a migration
+	Run() error
+}
+
+var _ Migration = &ToVersion1_1{}
+
+// ToVersion1_1 is an object responsible for migrating a fileset to version 1.1
+type ToVersion1_1 struct {
+	newMergerFn fs.NewMergerFn
+	infoFile    fs.ReadInfoFileResult
+	shard       uint32
+	namespace   namespace.Metadata
+	opts        storage.Options
+}
+
+func NewToVersion1_1(infoFile fs.ReadInfoFileResult, shard uint32, namespace namespace.Metadata, opts storage.Options) Migration {
+	return ToVersion1_1{
+		newMergerFn: fs.NewMerger,
+		infoFile:    infoFile,
+		shard:       shard,
+		namespace:   namespace,
+		opts:        opts,
+	}
+}
+
+// Run executes the steps to bring a fileset to Version 1.1
+func (v ToVersion1_1) Run() error {
+	reader, err := fs.NewReader(v.opts.BytesPool(), v.opts.CommitLogOptions().FilesystemOptions())
+	if err != nil {
+		return err
+	}
+
+	merger := v.newMergerFn(reader, v.opts.DatabaseBlockOptions().DatabaseBlockAllocSize(),
+		v.opts.SegmentReaderPool(), v.opts.MultiReaderIteratorPool(),
+		v.opts.IdentifierPool(), v.opts.EncoderPool(), v.opts.ContextPool(), v.namespace.Options())
+
+	volIndex := v.infoFile.Info.VolumeIndex
+	fsID := fs.FileSetFileIdentifier{
+		Namespace:   v.namespace.ID(),
+		Shard:       v.shard,
+		BlockStart:  xtime.FromNanoseconds(v.infoFile.Info.BlockStart),
+		VolumeIndex: volIndex,
+	}
+
+	nsCtx := namespace.NewContextFrom(v.namespace)
+
+	flushPersist, err := v.opts.PersistManager().StartFlushPersist()
```

Comment:
This seems to be the standard way to grab the persist manager. However, in `fs/source.go` (i.e. the filesystem bootstrap where this will be executed), it's using a `bootstrapper.SharedPersistManager` (https://github.com/m3db/m3/blob/master/src/dbnode/storage/bootstrap/bootstrapper/fs/source.go#L75) which wraps a persist manager with a mutex for shared use. Looking deeper, it appears that SharedPersistManager is used while writing the index (note: specifically mean the reverse index here and not index files) near the end of the bootstrapping process. Given that, is this safe for me to do when attempting to execute migrations in parallel? These migrations will block the bootstrap process until done (i.e. happens before the reverse index is built) and are operating on distinct filesets so seems like it should be OK?

Reply:
Yeah so - unfortunately you cannot use a persist manager concurrently since there is a single fileset writer. 

You'll probably have to create your own per concurrent "worker". Which means a worker pool is probably less than ideal since there's no concept of longly lived shared data structures when using the `sync.WorkerPool`.

Reply:
I don't think it's safe to use concurrently w/ other bootstrap processes. At the very minimum there are re-usable resources w/in the data persist manager that can't be shared.
```
type dataPersistManager struct {
        // Injected types.
        writer                        DataFileSetWriter
        nextSnapshotMetadataFileIndex nextSnapshotMetadataFileIndexFn
        snapshotMetadataWriter        SnapshotMetadataFileWriter

        // segmentHolder is a two-item slice that's reused to hold pointers to the
        // head and the tail of each segment so we don't need to allocate memory
        // and gc it shortly after.
        segmentHolder []checked.Bytes

        // The type of files that are being persisted. Assists with decision making
        // in the "done" phase.
        fileSetType persist.FileSetType

        // The ID of the snapshot being prepared. Only used when writing out snapshots.
        snapshotID uuid.UUID
}
```

```
 pm.dataPM.segmentHolder[0] = segment.Head
 pm.dataPM.segmentHolder[1] = segment.Tail
 err := pm.dataPM.writer.WriteAll(metadata, pm.dataPM.segmentHolder, checksum)
```

The `DataFileSetWriter` gets opened w/ specific options, I didn't look inside but it seems like it can't be shared concurrently either. 
```
dataWriterOpts := DataWriterOpenOptions{
        BlockSize: blockSize,
        Snapshot: DataWriterSnapshotOptions{
                SnapshotTime: snapshotTime,
                SnapshotID:   snapshotID,
        },
        FileSetType: opts.FileSetType,
        Identifier: FileSetFileIdentifier{
                Namespace:   nsID,
                Shard:       shard,
                BlockStart:  blockStart,
                VolumeIndex: volumeIndex,
        },
}
if err := pm.dataPM.writer.Open(dataWriterOpts); err != nil {
        return prepared, err
}
```